File: 백준/4256_트리/Main.java
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

class Solution {

    private int T;
    private int[] N;
    private List<Integer>[] preorders;
    private List<Integer>[] inorders;

    static class Node {
        int value;
        Node left;
        Node right;

        Node(int value) {
            this.value = value;
        }
    }

    public Solution(BufferedReader reader) throws IOException {
        T = Integer.parseInt(reader.readLine());
        N = new int[T];
        preorders = new ArrayList[T];
        inorders = new ArrayList[T];
        for(int i = 0 ; i < T ; i++) {
            N[i] = Integer.parseInt(reader.readLine());
            preorders[i] = Arrays.stream(reader.readLine().split(" "))
                    .map(Integer::parseInt)
                    .collect(Collectors.toList());
            inorders[i] = Arrays.stream(reader.readLine().split(" "))
                    .map(Integer::parseInt)
                    .collect(Collectors.toList());
        }
    }

    private Node recoverTree(List<Integer> preorder,
                             List<Integer> inorder) {
        if(preorder.size() == 0) {
            return null;
        }
        Node root = new Node(preorder.get(0));
//        System.out.println("root : " + root.value);
        int rootIndex = inorder.indexOf(root.value);

        List<Integer> leftInorder = inorder.subList(0, rootIndex);
        List<Integer> rightInorder = inorder.subList(rootIndex + 1, inorder.size());
        List<Integer> leftPreorder = preorder.subList(1, rootIndex + 1);
        List<Integer> rightPreorder = preorder.subList(rootIndex + 1, preorder.size());


        root.left = recoverTree(leftPreorder, leftInorder);
        root.right = recoverTree(rightPreorder, rightInorder);

        return root;
    }

    private void postOrder(Node root, StringBuilder sb) {
        if(root == null) {
            return;
        }
        postOrder(root.left, sb);
        postOrder(root.right, sb);
        sb.append(root.value).append(" ");
    }

    public void run() throws IOException {
        StringBuilder sb = new StringBuilder();
        // 전위 순회 결과와 후위 순위 결과가 주어졌을 때, 원본 트리를 복구
        for(int i = 0 ; i < inorders.length ; i++) {
            Node tree = recoverTree(preorders[i], inorders[i]);
            postOrder(tree, sb);
            sb.append("\n");
        }
        System.out.println(sb);
    }
}

File: 백준/4256_트리/main.rs
ulse std::io;

trait Solution {
    fn run(& self);
}

struct Problem4256 {
    T: i32,
    node_counts: Vec<i32>,
    inorders: Vec<Vec<i32>>,
    preorders: Vec<Vec<i32>>
}

struct Node {
    value: i32,
    left:  Option<Box<Node>>,
    right: Option<Box<Node>>
}

impl Solution for Problem4256 {

    fn run(& self) {

        for i in 0..self.T {
            let preorder = &self.preorders[i as usize];
            let inorder = &self.inorders[i as usize];
            let root = self.recoverTree(preorder, inorder);
            self.postOrder(root);
            println!("");
        }
    }
}

impl Problem4256 {

    fn recoverTree<'a>(& self, preorder: &'a Vec<i32>, inorder: &'a Vec<i32>) -> Option<Box<Node>> {
        if preorder.is_empty() {
            return None
        }

        let root = preorder[0];
        let rootIndex = inorder.iter().position(|&x| x == root).unwrap();
        let leftInorder = &inorder[0..rootIndex as usize].to_vec();
        let rightInorder = &inorder[(rootIndex+1) as usize..inorder.len()].to_vec();
        let leftPreorder = &preorder[1..rootIndex+1 as usize].to_vec();
        let rightPreorder = &preorder[(rootIndex+1) as usize .. preorder.len()].to_vec();

        Some(Box::new(Node {
            value: root,
            left: self.recoverTree(leftPreorder, leftInorder),
            right: self.recoverTree(rightPreorder, rightInorder)
        }))
    }

    fn postOrder(& self, root: Option<Box<Node>>) {
        if let Some(node) = root {
            self.postOrder(node.left);
            self.postOrder(node.right);
            print!("{} ", node.value);
        }
    }
}

fn main() {
    let mut buf: String = String::new();
    let T: i32 = read_line().parse().unwrap();
    let mut nodes: Vec<i32> = Vec::new(); // vec![0; T as usize];
    let mut preorders: Vec<Vec<i32>> = Vec::new();
    let mut inorders: Vec<Vec<i32>> = Vec::new();

    for i in 0..T {
        let N: i32 = read_line().parse().unwrap();
        let preorder = read_line().split_whitespace().map(|x| x.parse().unwrap()).collect();
        let inorder = read_line().split_whitespace().map(|x| x.parse().unwrap()).collect();
        nodes.push(N);
        preorders.push(preorder);
        inorders.push(inorder);
    }

    let problem = Problem4256{ T: T, node_counts: nodes, inorders: inorders, preorders: preorders };
    problem.run();
}

fn read_line() -> String {
    let mut buf = String::new();
    std::io::stdin().read_line(&mut buf).unwrap();
    buf.trim().to_string()
}
